Recording of videos in offscreen canvas to control resolution

AlbertoCasasOrtiz opened the pull request:
Hi @aymanhab, I am doing the recording of mp4 and mov videos in an offscreen canvas to control resolution and quality (to allow 4k). This is still a work in progress with not frontend to modify settings yet, but its working by changing them manually in the code.

I am trying to follow how you did it for screenshots, allowing to select default size (as seen in screen) or set custom.

I have a few topics for discussion here:
 - If screen aspect ratio is not preserved, the image can be highly distorted (this happens in screenshots too). Should we allow only custom resolution with fixed aspect ratio?
 - Related to this, I noticed allowing custom resolutions as numbers can be confusing to some users. Maybe it would be better to allow selecting resolutions in the same way as Youtube and many video editing software, which provides a selector with options (144p, 240p, 360p, 480p, 720p, 1080p_HD, 1440p_HD, 2160p_4K).
 - Since rendering offscreen takes time, higher resolutions can ralentize the application, and thus, affect framerate. I was thinking that instead of clicking play when animation recording starts, we should directly update frames from the recorder, so the recorder controls the animation in a way we ensure a fixed framerate. Again, I think providing a list of framerates could be more user friendly (15fps, 24fps, 30fps, 60fps) than allowing arbitrary framerates.

AlbertoCasasOrtiz committed:
Fixed background mismatch when transparency.

AlbertoCasasOrtiz committed:
Prepared for custom resolution recording.

aymanhab commented:
@AlbertoCasasOrtiz Since this is marked as WIP I assume no actual review is needed now. As to your questions:
1. Yes, distortions are possible, I wouldn't allow it for videos, rationale for snapshots is if publication needs a specific resolution/size users can get idea how that looks then tune, not sure how useful but some e.g. https://3dviewer.net/ expose this.
2. Perfectly fine with giving a dropdown instead of custom and see if users ask for more options
3.  I totally agree, we should create the frames at the speed of rendering which could be very slow for complex scenes. I don't know how you're testing this functionality but within the GUI context we probably need to control how often to send frames, there's some code to handle this. I can point you to it.

aymanhab commented:
@AlbertoCasasOrtiz Please let me know where this stands or how to help, ideally get this into the next beta. Thank you

AlbertoCasasOrtiz commented:
@aymanhab I have a working version of this, but with two caveats:

1. Since, in the GUI, from the viewer perspective I have not the movement animation available, I do not have control over it, so I cannot start the movement animation from the viewer when "record" is clicked (It works well with dollies).
2. Looks like it is glitching with high resolutions (4K), and I think it is because of mobx not updating frames in real time. We can discuss further this week.

My current approach is:
1. Record is clicked -> The animation is restarted.
2. For each frame:
   2.1. Render the frame offscreen at required resolution.
   2.2. Save that frame.
   2.3. Advance a frame.
3. Last frame is reached -> Stitch all frames together.

If we have control over frames, and we can interpolate frames as we require, we can have any FPS that we want. For now to keep it simple, I am keeping the original animation's framerate.

AlbertoCasasOrtiz committed:
Recording at specific resolutions.

AlbertoCasasOrtiz commented:
Hi @aymanhab 

This is ready to review. Glitches during recording are fixed in both GUI and viewer. Only two things pending that I think should be assessed as separate issues:

1. Sending a signal from viewer to GUI when record starts so the motion animation starts too.
2. Assess fps properly.
3. Assess white bands at right of video.

AlbertoCasasOrtiz committed:
Merge branch 'dev' into viewer_ui_tweaks

aymanhab commented:
@AlbertoCasasOrtiz is this ready for review? I'm worried relying on frame number will not work in GUI mode since we don't have frame numbers except for the dollies, as such I'm not sure why OpenSimGUIScene was changed at all?

AlbertoCasasOrtiz commented:
Hi @aymanhab. I think apart from the issues I mentioned in my previous comment this is ready for review, the issues are:

- Sending a signal from viewer to GUI when record starts so the motion animation starts too.
- Assess fps properly (I am currently working on this, but will address in a different PR).
- Assess white bands at right of video.

> I'm worried relying on frame number will not work in GUI mode since we don't have frame numbers except for the dollies, as such I'm not sure why OpenSimGUIScene was changed at all?

I found that, at least for dollies, we can use decimal frame numbers to control speed and framerate and I am using that to control fps in the generated video.

The reason to change OpenSimGUIScene was because if we control either time or frame number from the recorder while recording, there was a race condition even for the dollies. The changes I introduced avoid that race condition to happen.

AlbertoCasasOrtiz committed:
Added fps for videos. Time slider as time instead of frames.

AlbertoCasasOrtiz committed:
Recording at specified aspect ratios.

AlbertoCasasOrtiz committed:
Added snackbar message.

AlbertoCasasOrtiz committed:
Removed webm.

AlbertoCasasOrtiz commented:
@aymanhab This is ready to review, I fixed fps and aspect ratio errors and now you can record videos at any desired quality and specified aspect ratio and fps. I also removed lower quality options, 120 fps and webm as discussed in demos.

Only remaining issue related to this is:

- Sending a signal from viewer to GUI when record starts so the motion animation starts too.

aymanhab committed:
Send messages to the GUI to start/stop animation

aymanhab committed:
Merge branch 'viewer_ui_tweaks' of https://github.com/opensim-org/opensim-viewer into viewer_ui_tweaks

aymanhab commented:
@AlbertoCasasOrtiz I added a call to send message to the GUI to start playing animation, while not implemented on the GUI side yet there're wrinkles that need to be sorted out with (time/frame-number), handling the case where there's no current animation in the GUI or even if there's current animation and the user wants to navigate and record. We need to discuss but that's the reason I didn't merge this so as not to break existing/expected GUI functionality.

aymanhab committed:
Merge remote-tracking branch 'remotes/origin/dev' into viewer_ui_tweaks

aymanhab committed:
Handle animation in GUI, sending and receiving messages.

aymanhab committed:
Acccount for animation speed in the GUI to compute expected number of frames
